src/hotspot/cpu/riscv/gc/shared/barrierSetAssembler_riscv.cpp

```diff
@@ -231,13 +231,19 @@ void BarrierSetAssembler::incr_allocated_bytes(MacroAssembler* masm,
   __ sd(tmp1, Address(xthread, in_bytes(JavaThread::allocated_bytes_offset())));
 }
 
+extern int nmethod_barrier_guard_offset();
+
 void BarrierSetAssembler::nmethod_entry_barrier(MacroAssembler* masm) {
   BarrierSetNMethod* bs_nm = BarrierSet::barrier_set()->barrier_set_nmethod();
 
   if (bs_nm == NULL) {
     return;
   }
 
+  // RVC: RISCV's amoswap instructions need an alignment for the memory address it swaps
+  //   when we reach here we may get a 2-byte alignment so need to align it
+  __ align(4, nmethod_barrier_guard_offset());
```

Comment:
Is it necessary to pass nmethod_barrier_guard_offset() as the "extra_offset" parameter here since this function will return (12 * instruction_size) which is multiple of 4 ?

Reply:
Yes - it seems a little obscure though.

1. The background is that:

about the `nmethod_entry_barrier`:
```
nmethod_entry_barrier (begin) {   <-  when we reach here, we may get a 2-byte alignment because of RVC's existence.
   ... 
   amoswap                            
   <the swapped address in the code segment>    <-  though the size of the whole stub is a multiple of 4 currently, 
                                                    here we may also get a 2-byte alignment because of the above issue.
} nmethod_entry_barrier (end)
```

and about the trampoline:
```
trampoline (begin) {  <-  when we reach here, we may get a 2-byte alignment because of RVC's existence.
   auipc
   ld
   jalr
   <the real jump target in the code segment>  <-  same as the above explanation.
} trampline (end)
```

2. The reason to pass another argument is that: 

we might consider just keeping the start address of the code slices (nmethod_entry_barrier and trampoline) aligned to 4-byte to solve this issue, but this assumption is based on 'the size of the code slices (nmethod_entry_barrier and trampoline themselves) are a multiple of 4', and we don't know if they will change in the future by RVC. So I explicitly pass the real target address to the `align` function as a new argument to make sure the code is always right in the future.

I am pleased with other suggestions for this part.

Reply:
For the nmethod_entry_barrier case, I would omit the extra_offset parameter (and also the nmethod_barrier_guard_offset funtion). I think your newly-added assertion at line 265 is enough.

But looks like it's different in the trampoline case where I see we need a wordSize (8bytes) alignment. So we may keep the change for this case.

Reply:
Yes, I agree with your point. That trampoline part is special and requires to be scrupulous to make changes.

Reply:
Yes. Please also update the code comments. It looks strange to find those comments starts with a "RVC:".
Also we should make the exact alignment required explicit in comments like: "RISCV's amoswap instructions need an alignment for the memory address it swaps" or "RISCV CAS needs an alignment for memory", etc.

Reply:
done - polished and it needs some testing works.